### PR TITLE
[Embedding] Add inf-cl in embedding trainer

### PR DESCRIPTION
**Pull request by jie-z-0607**
### PR types

Function optimization
### PR changes

Others
### 经测试，inf-cl算子能够与原有损失函数有效对齐：
* 以数据类型设置bf16，group_size设置1，gradient_accumulation_steps设置4为例，inf_cl_loss与原有contrastive_loss的收敛曲线如下：
image
### 经测试，在超大batch_size下，inf-cl算子能够有效降低embedding训练时的显存消耗：
* 在8张A100（80G）显卡下，以数据类型设置bf16，group_size设置4，gradient_accumulation_steps设置4096为例，inf_cl_loss与原有contrastive_loss的显存占用对比如下：

<meta charset="UTF-8"><meta charset="UTF-8"><div class="mp-table-align"><div class="mp-table-wrapper"><div data-slate-node="element" style="padding-left:0px" class="mp-table-container">
参数设置 | 显存占用 | 首个step完成耗费时间
-- | -- | --
不使用inf-cl；embedding_negatives_cross_device=True | 42238MiB；42526MiB；<br>42526MiB；42470MiB；<br>42470MiB；42526MiB；<br>42526MiB；42182MiB | 48min42s
使用inf-cl；embedding_negatives_cross_device=Flase | 29630MiB；28392MiB；<br>28372MiB；28308MiB；<br>28320MiB；28384MiB；<br>28316MiB；28070MiB | 49min56s

</div><div class="mp-table-serialize-flag"><br/></div></div></div><span data-morpho-doc-data='{"token":"eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwiYXBwSWQiOjEsInVpZCI6InlERTQxZTd2N2YiLCJkb2NJZCI6IjFhcWE5MGxKXzNwTElRIn0..ENewSAzvjpZ2z_XV.ezPQiTtPKQ0FoRzCzVvPZl4zb7q2IDxI5IXSNu0NfNpkBFy3JvYtMa4WBzlA-CLqn8L7Ob4xMcmGsD3OCRBYYH4GGCm9wJeNyb8PbJPHZJ1QW3qGUUKey5cne6PMtivHdXUla7Ooso4CDUOgb101YO_2fO4hpfvSXZo_pwfMWqBCOVZlo4WDh7w-vnszmWFr_yNwW_4A0Gs4RaWqcDikClNy2A.nTAXLrgFlzthGxrj9Qn67w","appId":"1"}' class='mp-morpho-clipboard-doc-data'>﻿</span>

* 在8张A100（80G）显卡下，以数据类型设置bf16，group_size设置1，gradient_accumulation_steps设置16384（总计batch_size 128K）为例，inf_cl_loss与原有contrastive_loss的显存占用对比如下：

<meta charset="UTF-8"><meta charset="UTF-8"><div class="mp-table-align"><div class="mp-table-wrapper"><div data-slate-node="element" style="padding-left:0px" class="mp-table-container">
参数设置 | 显存占用 | 首个step完成耗费时间
-- | -- | --
不使用inf-cl；embedding_negatives_cross_device=True | 超出显存限制 |  
使用inf-cl；embedding_negatives_cross_device=Flase | 46324MiB；45192MiB；<br>44926MiB；45180MiB；<br>44674MiB；45022MiB；<br>45032MiB；44904MiB | 2h23min46s

</div><div class="mp-table-serialize-flag"><br/></div></div></div><span data-morpho-doc-data='{"token":"eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwiYXBwSWQiOjEsInVpZCI6InlERTQxZTd2N2YiLCJkb2NJZCI6IjFhcWE5MGxKXzNwTElRIn0..ENewSAzvjpZ2z_XV.ezPQiTtPKQ0FoRzCzVvPZl4zb7q2IDxI5IXSNu0NfNpkBFy3JvYtMa4WBzlA-CLqn8L7Ob4xMcmGsD3OCRBYYH4GGCm9wJeNyb8PbJPHZJ1QW3qGUUKey5cne6PMtivHdXUla7Ooso4CDUOgb101YO_2fO4hpfvSXZo_pwfMWqBCOVZlo4WDh7w-vnszmWFr_yNwW_4A0Gs4RaWqcDikClNy2A.nTAXLrgFlzthGxrj9Qn67w","appId":"1"}' class='mp-morpho-clipboard-doc-data'>﻿</span>

### CODE REVIEW

File: ops/src/paddlenlp_kernel/triton/inf_cl/inf_cl_loss.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import paddle
+import paddle.nn as nn
+
+from .ring import cal_inf_loss
+
+__all__ = ["Simple_Inf_cl_loss", "Matryoshka_Inf_cl_loss"]
+
+
+class Simple_Inf_cl_loss(nn.Layer):
```
Comment (ZHUI) [inline ops/src/paddlenlp_kernel/triton/inf_cl/inf_cl_loss.py]: 加一些注释

File: paddlenlp/trl/embedding_trainer.py [hunk 2]
```diff
@@ -18,6 +18,10 @@
 from paddle.base import core
 from paddle.distributed import fleet
 
+from ops.src.paddlenlp_kernel.triton.inf_cl.inf_cl_loss import (
```
Comment (ZHUI) [inline paddlenlp/trl/embedding_trainer.py]:
```suggestion
from paddlenlp_kernel.triton.inf_cl.inf_cl_loss import (
```

Comment (ZHUI) [inline paddlenlp/trl/embedding_trainer.py]: 这个没有默认安装,需要 try except一下

File: paddlenlp/transformers/contrastive_loss.py [hunk 3]
```diff
@@ -63,3 +72,84 @@ def forward(self, q_reps, p_reps):
...
+            q_reps (Tensor): Query representations.
+            p_reps (Tensor): key representations.
+
+        Returns:
+            Tensor: The computed loss.
+        """
+        group_size = p_reps.shape[0] // q_reps.shape[0]  # Number of keys per query
+        labels = paddle.arange(q_reps.shape[0], dtype="int64")  # Generate labels for queries
+        labels = labels * group_size  # Adjust labels based on group size
+        loss = cal_inf_loss(q_reps, p_reps, labels=labels, scale=None, head_dim=self.head_dim)
```
Comment (ZHUI) [inline paddlenlp/transformers/contrastive_loss.py:99]: 你把import 的代码放到这里吧, 然后没有包的话，直接报错。
```

try:
    from paddlenlp_kernel.triton.inf_cl import cal_inf_loss
except ImportError:
    logger.warning(
        "Paddlenlp_kernels are not available, which means the inf_cl loss cannot be used. If you wish to use the inf_cl loss, please follow the instructions in the README.md on the `ops`."
    )
```

**Review by DesmonDay — APPROVED** [review]
